src/ratbag_emu/protocol/util/profile.py

```diff
@@ -35,8 +41,7 @@ def __init__(self, obj=None):
                     self.x_dpi[i] = getattr(res, 'xres')
                     self.y_dpi[i] = getattr(res, 'yres')
 
-                    if hasattr(res, 'is_active'):
-                        if getattr(res, 'is_active'):
+                    if hasattr(res, 'is_active') and getattr(res, 'is_active'):
                             self.active_dpi = i
```

Comment:
It is actually

```python

try:
	if res['is_active']:
		self.active_dpi = i
except AttributeError:
	pass
```

which I think is way worse than

```python
if hasattr(res, 'is_active') and getattr(res, 'is_active'):
	self.active_dpi = i
```

Reply:
rule of thumb: *never* use `getattr`, `hasattr` and `setattr` unless you are doing introspection, which doesn't seem to be the case here.
So I tend to disagree:
```python
try:
    if res.is_active:
        self.active_dpi = res
except AttributeError:
    pass
```
Is the best pythonic way of sorting this out.

Note that I intentionally used `res` in the assignment, as why keeping the dpi value when you can keep a reference to the object itself.

second rule of thumb: `AttributeError` means a poorly written class, because if you intend to access at some point one attribute, you *must* initialize it to not cripple you code with `try/catch` everywhere.

So, after the second rule, the best way is:
```python
 if res.is_active:
    self.active_dpi = res
```

Reply:
> second rule of thumb: AttributeError means a poorly written class, because if you intend to access at some point one attribute, you must initialize it to not cripple you code with try/catch everywhere.

This value doesn't come from a class, it comes from the json data we receive.

I will use the try block.

Reply:
> This value doesn't come from a class, it comes from the json data we receive.

There is something odd here. IIRC, when you parse a json object, you get a plain dict or list.
But `hasattr`, `getattr` and `setattr` should not work on those, especially in the way you are using them.

FWIW, doing `getattr(a, 'constant')` is strictly equivalent to `a.constant`.

So I maintain that you have an object somewhere.

If you really do have a json object, you should use (if you want to convert the `hasattr`:
```python
if 'is_active' in res and res['is_active']:
    self.active_dpi = res
```

but doing the try catch is better, in the python world:
```python
try:
    if res['is_active']:
        self.active_dpi = res
except KeyError:
    pass
```

One last thing: `obj`should probably be renamed `json_object`, and IMO it would be best to convert the json data you get first into a proper python object where you ensure you have your proper API:
```python
class MyObject:
  def __init__(self):
    self.is_active = False
    self.xres = 1000
    self.yres = 1000

  @classmethod
  from_json(cls, json_obj):
    returned_object = cls()
    for key in ['is_active', 'xres', 'yres']:
       if key in in json_obj:
         setattr(returned_object, key) = json_obj[key]

# then convert a json to MyObject:
new_obj = MyObject.from_json(json_data)
```

> I will use the try block.

thanks

Reply:
Doesn't work here but an FYI:  dictionary values can be retrieved with `get()` which defaults to `None` or the given value. So `mydict.get('somekey', 'defaultvalue')` returns either `mydict['somekey']` or `defaultvalue` where the key is not present. That way you don't have to deal with exceptions.